src/mlpack/methods/decision_tree/decision_tree_regressor_impl.hpp

```diff
@@ -0,0 +1,1081 @@
+        typename std::remove_reference<
+        WeightsType>::type>::value>*)
+{
+  // Sanity check on data.
+  util::CheckSameSizes(data, labels, "DecisionTreeRegressor::Train()");
+
+  using TrueMatType = typename std::decay<MatType>::type;
+  using TrueLabelsType = typename std::decay<LabelsType>::type;
+  using TrueWeightsType = typename std::decay<WeightsType>::type;
+
+  // Copy or move data.
+  TrueMatType tmpData(std::move(data));
+  TrueLabelsType tmpLabels(std::move(labels));
+  TrueWeightsType tmpWeights(std::move(weights));
+
+  // Set the correct dimensionality for the dimension selector.
+  dimensionSelector.Dimensions() = tmpData.n_rows;
+
+  // Pass off work to the Train() method.
+  return Train<true>(tmpData, 0, tmpData.n_cols, tmpLabels, 0,
+      tmpWeights, minimumLeafSize, minimumGainSplit, maximumDepth,
+      dimensionSelector);
+}
+
+//! Train on the given data.
+template<typename FitnessFunction,
+         template<typename> class NumericSplitType,
+         template<typename> class CategoricalSplitType,
+         typename DimensionSelectionType,
+         bool NoRecursion>
+template<bool UseWeights, typename MatType, typename LabelsType>
+double DecisionTreeRegressor<FitnessFunction,
+                             NumericSplitType,
+                             CategoricalSplitType,
+                             DimensionSelectionType,
+                             NoRecursion>::Train(
+    MatType& data,
+    const size_t begin,
+    const size_t count,
+    const data::DatasetInfo& datasetInfo,
+    LabelsType& labels,
+    const size_t numClasses,
+    arma::rowvec& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    const size_t maximumDepth,
+    DimensionSelectionType& dimensionSelector)
+{
+  // Clear children if needed.
+  for (size_t i = 0; i < children.size(); ++i)
+    delete children[i];
+  children.clear();
+
+  // Look through the list of dimensions and obtain the gain of the best split.
+  // We'll cache the best numeric and categorical split auxiliary information in
+  // numericAux and categoricalAux (and clear them later if we make no split),
+  double bestGain = FitnessFunction::template Evaluate<UseWeights>(
+      labels.subvec(begin, begin + count - 1),
+      numClasses,
+      UseWeights ? weights.subvec(begin, begin + count - 1) : weights);
+  size_t bestDim = datasetInfo.Dimensionality(); // This means "no split".
+  const size_t end = dimensionSelector.End();
+
+  if (maximumDepth != 1)
+  {
+    for (size_t i = dimensionSelector.Begin(); i != end;
+         i = dimensionSelector.Next())
+    {
+      double dimGain = -DBL_MAX;
+      if (datasetInfo.Type(i) == data::Datatype::categorical)
+      {
+        dimGain = CategoricalSplit::template SplitIfBetter<UseWeights>(bestGain,
+            data.cols(begin, begin + count - 1).row(i),
+            datasetInfo.NumMappings(i),
+            labels.subvec(begin, begin + count - 1),
+            numClasses,
```

Comment:
Yeah, we can do that. I kept it that way because I wanted to reuse code and I simply adapted the old AllCategoricalSplit that used this parameter. It will be easy to remove this. So, I think we can do this at the end. The thing is there is still some confusion about how we are storing info in the splitting strategies. So, I think once that part is sorted, we can remove this easily if needed.